include/aws/common/byte_buf.h

```diff
@@ -35,5 +39,140 @@ static inline struct aws_byte_buf aws_byte_buf_from_array(const uint8_t c_str[],
+ * within [0, bound) even under CPU speculation conditions, and is intended to
+ * be used for SPECTRE mitigation purposes.
+ */
+static inline size_t aws_nospec_index(size_t index, size_t bound) {
+    /*
+     * SPECTRE mitigation - we compute a mask that will be zero if len < 0
+     * or len >= buf->len, and all-ones otherwise, and AND it into the index.
+     * It is critical that we avoid any branches in this logic.
+     */
+
+    /*
+     * Hide the index value from the optimizer. This helps ensure that all this
+     * logic doesn't get eliminated.
+     */
+#if defined(__GNUC__) || defined(__clang__)
+     __asm__ __volatile__("" : "+r" (index));
+#endif
+#if defined(_MSVC_LANG)
+     /*
+      * MSVC doesn't have a good way for us to blind the optimizer, and doesn't
+      * even have inline asm on x64. Some experimentation indicates that this
+      * hack seems to confuse it sufficiently for our needs.
+      */
+     *((volatile uint8_t *)&index) += 0;
+#endif
+
+    /*
+     * If len > (SIZE_MAX / 2), then we can end up with len - buf->len being
+     * positive simply because the sign bit got inverted away. So we also check
+     * that the sign bit isn't set from the start.
+     *
+     * We also check that bound <= (SIZE_MAX / 2) to catch cases where the buffer
+     * is _already_ out of bounds.
+     */
+    size_t negative_mask = index | bound;
+    size_t toobig_mask = bound - index - (uintptr_t)1;
+    size_t combined_mask = negative_mask | toobig_mask;
+
+    /*
+     * combined_mask needs to have its sign bit OFF for us to be in range.
+     * We'd like to expand this to a mask we can AND into our index, so flip
+     * that bit (and everything else), shift it over so it's the only bit in the
+     * ones position, and multiply across the entire register.
+     *
+     * First, extract the (inverse) top bit and move it to the lowest bit.
+     * Because there's no standard SIZE_BIT in C99, we'll divide by a mask with
+     * just the top bit set instead.
+     */
+    
+    combined_mask = (~combined_mask) / (SIZE_MAX - (SIZE_MAX >> 1));
+
+    /* 
+     * Now multiply it to replicate it across all bits.
+     *
+     * Note that GCC is smart enough to optimize the divide-and-multiply into
+     * an arithmetic right shift operation on x86.
+     */
+    combined_mask = combined_mask * SIZE_MAX;
+
+    return index & combined_mask;
+}
+
+/**
+ * Tests if the given aws_byte_cursor has at least len bytes remaining. If so,
+ * *buf is advanced by len bytes (incrementing ->ptr and decrementing ->len),
+ * and an aws_byte_cursor referring to the first len bytes of the original *buf is
+ * returned. Otherwise, an aws_byte_cursor with ->ptr = NULL, ->len = 0 is returned.
+ *
+ * Note that if len is above (SIZE_MAX / 2), this function will also treat it as a buffer
+ * overflow, and return NULL without changing *buf.
+ */
+static inline struct aws_byte_cursor aws_byte_cursor_advance(struct aws_byte_cursor *cursor, size_t len) {
```

Comment:
This construction is a little bit confusing to me, though admittedly my opinion may be swayed by the way you were using the cursor construct in the aws-encryption-sdk-c code. But if we've already checked that the original buffer we were inspecting had at least len bytes, I don't see what the advantage of returning a new struct aws_byte_cursor with len bytes is. The only thing that cursor can be used for is a redundant length check.